### PR TITLE
Implement PWM Driver for Sonata Board

### PR DESCRIPTION
**Pull request by AlexJones0**
This PR implements a driver for the Pulse-Width Modulation (PWM) used on the Sonata board.

Note that the linked documentation source is currently out of date - I believe that @HU90m intends to update that to reflect the actual hardware, which is what this driver is based on.

I've confirmed that this is working for the 1 PWM that we have on Sonata by implementing the following in a simple piece of firmware with a single compartment: 
```cpp
#include <compartment.h>
#include <platform-pwm.hh>
#include <thread.h>

void __cheri_compartment("pwm_test") pwm_test()
{	
	auto pwm = MMIO_CAPABILITY(SonataPulseWidthModulation, pwm);
	uint8_t counter = 0;
	while (true) 
        {
		counter = (counter + 1) % 255;
		pwm->output_set(0, 255, counter);
		thread_millisecond_wait(50);
	}
}
```

Checking with a Multimeter I can then see analogue variation between 0V and 3.3V on the mikroBUS PWM output. 

Sonata currently only supports 1 PWM but this could potentially be changed, so I've designed the driver to be extensible for this case.

### CODE REVIEW

File: sdk/include/platform/sunburst/platform-pwm.hh
Comment (HU90m) [inline sdk/include/platform/sunburst/platform-pwm.hh:58]: Since 255 is the maximum PWM period, I think it would be nicer to have `period` and `dutyCycle` be `uint8_t` and remove the `MaxPwmPeriod`. We don't need this to be parameterised.

File: sdk/include/platform/sunburst/platform-pwm.hh
Comment (HU90m) [inline sdk/include/platform/sunburst/platform-pwm.hh:27]:
```suggestion
	 * The number of PWM outputs available.
```

File: sdk/include/platform/sunburst/platform-pwm.hh
Comment (HU90m) [inline sdk/include/platform/sunburst/platform-pwm.hh:49]: A little more here would be nice. Something like: "The period of the output, set with the number of clock cycles a period should last."

File: sdk/include/platform/sunburst/platform-pwm.hh
Comment (HU90m) [inline sdk/include/platform/sunburst/platform-pwm.hh:55]: Please could you document the parameters of the function.

**Review by HU90m — CHANGES_REQUESTED** [review]

File: sdk/include/platform/sunburst/platform-pwm.hh
Comment (AlexJones0) [reply sdk/include/platform/sunburst/platform-pwm.hh:58]: Ok, in that case since the Max PWM period is unlikely to change I'll remove the static variable and make the `period` and `dutyCycle` a `uint8_t`, and I'll just document the counter size somewhere so that it is clear when using the interface why a `uint8_t` is supplied if the register is a `uint32_t`.

File: sdk/include/platform/sunburst/platform-pwm.hh
Comment (AlexJones0) [reply sdk/include/platform/sunburst/platform-pwm.hh:27]: Agreed - also changing the name `pwmCount` to `pwmOutputCount` so this is more explicit that we are talking about possible PWM outputs and not individual devices.

File: sdk/include/platform/sunburst/platform-pwm.hh
Comment (AlexJones0) [reply sdk/include/platform/sunburst/platform-pwm.hh:19]: Sure - in this case I was making the debug static variables follow the naming convention of the other Sonata drivers, e.g. `debugEthernet` or `debugOpenTitanI2c` or `debugSonataSpi`, but if you think that it's not worth it it here then I'm happy to change this.

File: sdk/include/platform/sunburst/platform-pwm.hh
Comment (AlexJones0) [reply sdk/include/platform/sunburst/platform-pwm.hh:6]: Understood - I intend to maintain the `Pwm` in the struct name so that we only have to use `SonataPwm` instead of `SonataPulseWidthModulation` every time we want to instantiate PWM - be sure to let me know if you want this changed as well. Instead, I've updated the high-level documentation to expand the abbreviation so that it is clear what it is referring to, and have mostly scrubbed references to the abbreviation from other parts of the code.

File: sdk/include/platform/sunburst/platform-pwm.hh
Comment (davidchisnall) [reply sdk/include/platform/sunburst/platform-pwm.hh:19]: Probably fine for the debug value, but not so much for the public APIs.

File: sdk/include/platform/sunburst/platform-pwm.hh
Comment (davidchisnall) [reply sdk/include/platform/sunburst/platform-pwm.hh]: No, please use `SonataPulseWidthModulation`.  If someone wants to use a shorter local name, they can use a `using` directive.

Comment (AlexJones0) [reply sdk/include/platform/sunburst/platform-pwm.hh]: Understood, I'll make that change now.

**Comment by AlexJones0** [comment]
Thanks for the comments - I've updated the driver to address all the comments as far as I can tell, and I've updated the test code to fit the new interface. I can confirm again that based on the simple test the driver is working by observing the analogue output vary via a multimeter.

File: sdk/boards/sonata-prerelease.json
Comment (davidchisnall) [inline sdk/boards/sonata-prerelease.json:67]: @HU90m Is this also present in the 0.2 release?

**Review by davidchisnall — COMMENTED** [review]
LGTM, I'm happy if @HU90m is happy.

File: sdk/boards/sonata-prerelease.json
Comment (AlexJones0) [reply sdk/boards/sonata-prerelease.json:67]: > @HU90m Is this also present in the 0.2 release?

I've just checked for @HU90m, and while the PWM existed in the code at the point of the 0.2 release, it wasn't hooked up to any pins in the RTL and so there's no way to use it. I think in this case it doesn't make sense to add this to the 0.2 board file also.

**Review by HU90m — APPROVED** [review]
@HU90m is happy

**Comment by HU90m** [comment]
@davidchisnall this is awaiting your approval before it will merge

**Review by davidchisnall — APPROVED** [review]

**Pull request merged**
